src/Microsoft.AspNet.Server.Kestrel/Http/Frame.cs

```diff
@@ -596,7 +596,12 @@ private bool TakeStartLine(SocketInput input)
                 {
                     return false;
                 }
-                var requestUri = begin.GetString(scan);
+
+                var percent = begin;
+                var needDecode = percent.Seek('%') >= 0;
```

Comment:
This should be after the query handling, you don't care if there are % in the query.